SFS_general/delayline.m

```diff
@@ -1,149 +1,221 @@
+%                                                                            *
+% The above copyright notice and this permission notice shall be included in *
+% all copies or substantial portions of the Software.                        *
+%                                                                            *
+% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR *
+% IMPLIED, INCLUDING BUT  NOT LIMITED TO THE  WARRANTIES OF MERCHANTABILITY, *
+% FITNESS  FOR A PARTICULAR  PURPOSE AND  NONINFRINGEMENT. IN NO EVENT SHALL *
+% THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER *
+% LIABILITY, WHETHER  IN AN  ACTION OF CONTRACT, TORT  OR OTHERWISE, ARISING *
+% FROM,  OUT OF  OR IN  CONNECTION  WITH THE  SOFTWARE OR  THE USE  OR OTHER *
+% DEALINGS IN THE SOFTWARE.                                                  *
+%                                                                            *
+% The SFS Toolbox  allows to simulate and  investigate sound field synthesis *
+% methods like wave field synthesis or higher order ambisonics.              *
+%                                                                            *
+% http://sfstoolbox.org                                 sfstoolbox@gmail.com *
+%*****************************************************************************
+
+
+%% ===== Configuration ===================================================
+fracdelay = conf.fracdelay;
+
+
+%% ===== Computation =====================================================
+% --- Reshape signals ---
+% Check if the impulse response is given in SOFA conventions [M C N], or in
+% usual [N C] convention, where
+% M ... number of measurements
+% C ... number of channels
+% N ... number of samples
+if ndims(sig)==3
+    [M, C, samples] = size(sig);
+    channels = M * C;
+    % Reshape [M C N] => [N C*M], this will be redone at the end of the function
+    sig = reshape(sig,[channels,samples])';
+    reshaped = true;
+else
+    % Assume standard format [N C]
+    [samples, channels] = size(sig);
+    reshaped = false;
+end
+
+% If only single valued time delay and weight is given, create vectors
+if channels>1 && length(dt)==1, dt=repmat(dt,[1 channels]); end
+if channels>1 && length(weight)==1, weight=repmat(weight,[1 channels]); end
+
+rfactor = 1.0;  % ratio of signal length and delayline length
+switch fracdelay.pre.method
+    case 'resample'
+        % === Resample ===================================================
+        % resample factor (1/stepsize of fractional delays)
+        rfactor = fracdelay.pre.resample.factor;
+        switch fracdelay.pre.resample.method
+            case 'matlab'
+                sig = resample(sig,rfactor,1);
+            case 'pm'
+                % === Parks-McClellan linear phase FIR filter ===
+                a = [1 1 0 0];
+                f = [0.0 0.9/rfactor 1/rfactor 1.0];
+                b = firpm(fracdelay.pre.resample.order,f,a);
```

Comment:
Yeah, but it's all for pretending to have it:
```Matlab
>> help firpm
error: help: the 'firpm' function belongs to the signal package from Octave Forge but
has not yet been implemented.
```

We could of course do nothing as Octave produces a meaningful error, but I like to give the users a hint what they have to change in order to make it work.